### ISSUE TITLE
0.13.7

### ISSUE THREAD

**Issue by hdgarrood** [purescript/purescript#3886]
@justinwoo [recently wrote about some changes he backported](https://github.com/justinwoo/my-blog-posts/blob/master/posts/2020-05-22-some-backported-purescript-changes-for-my-own-uses.md) in order to make use of some commits which are in master now, and I think the post suggests that an official 0.13.7 release is worth considering. The CBOR externs change in particular is very exciting; I hadn’t fully appreciated how big a deal that is.

As for 0.14.0: it is not that far away, but there are a few bugs in roles that we will want to fix first. I think an 0.13.7 release would take the pressure off a bit, so to speak, so that we wouldn’t feel the need to rush those fixes to get it out sooner, too.

I’m happy to do the remaining legwork to put together a release (especially since Justin has done a lot of it already anyway). Of the commits in master since 0.13.6, I think the following are candidates for backporting (because they are non-breaking changes which will be useful downstream):

- #3777 fix some bounds
- #3781 add `purs graph` for graphing module dependencies
- #3789 update cache-db.json on ide rebuilds - fixes a bug which could lead to things not being rebuilt when they should be
- #3799 tracks the cache-db.json timestamp in purs ide in order to reload ide state when necessary, and get rid of polling and file watching (the flags are still present but are ignored). @kritzcreek this can be considered non breaking, right?
- #3841 binary encoding for externs
- #3850 hide compiler internals in Prim modules from ide completions
- #3843 use a single Text for module names
- #3856 extract documentation for type classes in purs ide
- #3877 Fix CoreFn JSON version parsing 

My plan is to create a branch off 0.13.6, try to pull these commits in one by one, and see how far I get.

**Comment by kritzcreek** [comment]
Yeah I made the removed CLI flags issue deprecation warnings.

Thanks a lot for looking into this!

**Comment by joneshf** [comment]
I'm all for us finally making a release. But, I'm pretty much against this approach we're taking in this release. We're addressing the symptom, but not the problem.

The symptom is that there's useful features available that are ready to go, but we haven't had a release in 18 weeks. The problem is that we think major version releases are special, so we can take as long as necessary to make them.

We're not willing to make a release because new bugs keep popping up in some of the new features (roles, coercible, others?). A similar thing happened with [the 0.12.0 release](https://github.com/purescript/purescript/issues/3178) a few years ago. It was stymied on some kind of bug with proxy syntax and whatever. Holding up any release on a new feature that introduced a bug is our pattern, and it's the problem. If we cannot get over this mental block that we have, we're going to continue to have this issue again and again.

The whole motivation behind moving to time-based releases was to prevent the situation we got into in the past and we're in right now. The idea is to prevent us from saying, "just one more thing" for an indefinite period of time. The idea is to continue moving forward without getting bogged down on releasing bug-free software (as it's nigh impossible). The idea is to satiate people knowing when the next release is coming. The idea is to mitigate people having to fork the project just to get a non-breaking feature.

We're not doing time-based releases anymore because we're 18 weeks since the last release. We should have had three releases since then:
1. A 0.13.7 release the last week of February (2020-02-23 to 2020-02-29) for:
    * https://github.com/purescript/purescript/pull/3777
    * https://github.com/purescript/purescript/pull/3781
    * https://github.com/purescript/purescript/pull/3351 - This turned out to not be a breaking change.
    * https://github.com/purescript/purescript/pull/3774
    * https://github.com/purescript/purescript/pull/3789
1. A 0.14.0 release the second week of April (2020-04-05 to 2020-04-11) for:
    * https://github.com/purescript/purescript/pull/3795
    * https://github.com/purescript/purescript/pull/3799
    * https://github.com/purescript/purescript/pull/3793 - This is at least breaking for `purescript` the library, which necessitates a major version bump per PVP.
    * https://github.com/purescript/purescript/pull/3810
    * https://github.com/purescript/purescript/pull/3779 - I _think_ this is breaking for PureScript the language, but it's definitely breaking for `purescript` the library, which necessitates a major version bump per PVP.
    * https://github.com/purescript/purescript/pull/3811
    * https://github.com/purescript/purescript/pull/3817
    * https://github.com/purescript/purescript/pull/3821
    * https://github.com/purescript/purescript/pull/3808
    * https://github.com/purescript/purescript/pull/3823
    * https://github.com/purescript/purescript/pull/3792 - This is at least breaking for `purescript` the library, which necessitates a major version bump per PVP.
    * https://github.com/purescript/purescript/pull/3826
    * https://github.com/purescript/purescript/pull/3831
1. A 0.15.0 release this week (2020-05-17 to 2020-05-23) for:
    * https://github.com/purescript/purescript/pull/3841 - This is at least breaking for `purescript` the library, which necessitates a major version bump per PVP.
    * https://github.com/purescript/purescript/pull/3846
    * https://github.com/purescript/purescript/pull/3837
    * https://github.com/purescript/purescript/pull/3850
    * https://github.com/purescript/purescript/pull/3843 - Unclear if this is breaking.
    * https://github.com/purescript/purescript/pull/3853
    * https://github.com/purescript/purescript/pull/3829
    * https://github.com/purescript/purescript/pull/3847
    * https://github.com/purescript/purescript/pull/3856
    * https://github.com/purescript/purescript/pull/3839 - Pretty sure this is a breaking change since how `purs compile` works was changed.
    * https://github.com/purescript/purescript/pull/3816
    * https://github.com/purescript/purescript/pull/3864
    * https://github.com/purescript/purescript/pull/3861
    * https://github.com/purescript/purescript/pull/3877 - This is at least breaking for `purescript` the library, which necessitates a major version bump per PVP.

So that's everything that's supposed to be in this upcoming release, plus all the other stuff that people have contributed. It would have come gradually over the past 18 weeks. It could've fostered new tooling with the `purs graph` command being released 12 weeks ago. It could've fixed quite a few bugs with all that would've been released six weeks ago. More importantly, we could've mitigated someone wanting to fork the compiler just to get a useful feature that was already ready to go. If we don't want to do time-based releases anymore, that's fine. But, we need a better process than whatever we're doing now. It shouldn't take someone forking, making binaries, and blogging it for us to get motivated to make a release.

It's all well and good for us to make some release. We should celebrate it because releasing software is hard. But, there's important feedback in this whole situation: the release process for the compiler has problems and we're not fixing the problems.

**Comment by hdgarrood** [comment]
The approach that I thought we had agreed on was "aim for 6 weeks between releases, except for breaking releases, which may take longer." It sounds the process you're advocating is releasing every 6 weeks unconditionally, merging breaking changes whenever, and accept that a result of that is that breaking releases become more frequent? 

One problem with that approach is that stability (reducing the frequency and pain of breaking releases) is, in my impression, one of the things most desired by the community. The response to Phil's ["The state of things"](https://discourse.purescript.org/t/the-state-of-things/282) post on Discourse evidences that, for instance. It seems fairly clear to me that planning breaking changes more deliberately, as we have been doing for 0.14.0, allows us to reduce the frequency of breaking releases, which acts in service of this goal.

Another problem is that there's often a lot of things that need to happen outside of the compiler for us to be ready to make releases which include breaking changes; getting libraries updated is the main one. For the breaking changes which are currently in master, a lot of that work has been done already, but not all of it: the prelude emits warnings with the current compiler master because of missing kind annotations, and also we would need to add role declarations in some cases to prevent invalid coercions; Map needs its first type parameter to be `nominal`, Set needs its parameter to be `nominal` too, and there may be others. (Incidentally, this is why Coercible really needs to be in a breaking release: if we release it in a non-breaking release then people will be able to use it with current versions of core libraries, which don't have the necessary role declarations, and this permits invalid coercions. We can't add those role declarations without a major bump in the relevant libraries.)

We _could_ start saying we will continue to make releases every 6 weeks regardless of whether a breaking change is coming up, and branching as necessary to keep non-breaking changes and breaking changes separate, if breaking changes are in master but not quite ready to be shipped. But of course that adds quite a lot of overhead. I don't think it would be a good tradeoff as things stand currently.

**Comment by hdgarrood** [comment]
I managed to get all of the above in. v0.13.8 is now out: https://github.com/purescript/purescript/releases/tag/v0.13.8 :tada:

**Comment by garyb** [comment]
> Incidentally, this is why Coercible really needs to be in a breaking release: if we release it in a non-breaking release then people will be able to use it with current versions of core libraries, which don't have the necessary role declarations, and this permits invalid coercions.

Uhh, that doesn't sound right. I thought in the absence of annotation we defaulted to `nominal`?

**Comment by hdgarrood** [comment]
We only default to nominal for `foreign import data` types, because we have no idea what their representations look like. For normal data types we infer roles using the algorithm in the safe zero cost coercions paper. I don’t think Coercible would be very useful if we assumed that everything was nominal, because then almost every data type would need an explicit role annotation to permit coercions (and in practice lots of them probably wouldn’t get them). It is definitely unfortunate that types like Map where instance behaviour affects the representation are unsafe by default, but I don’t see a way around it that doesn’t make Coercible really awkward to use.

**Comment by garyb** [comment]
Ah okay, interesting. I assumed `Coercible` would be opt in, and I'd prefer that it was really - it seems kinda dangerous not to do so, and I'd take safety over convenience here.

I guess it's not dangerous if it's a fact everyone is aware of, but it certainly raises some issues retrofitting it into existing code, since I know of a whole bunch of types that preserve invariants other than `Map` and `Set`, that will also need the annotation adding. 

Getting a bit off track from this issue I suppose, but would it be possible to infer a nominal role for types that have private constructors? That would allay most of my concerns.

**Comment by natefaubion** [comment]
I’m not sure how much that would help. What if you export it from an internal module? There’s nothing “unsafe” about, it’s about invariants unavailable to the type system. That is, you won’t subvert type safety. I think it’s a fairly rare situation.

**Comment by hdgarrood** [comment]
Which other types will need role annotations?

**Comment by kl0tl** [comment]
@garyb `Coercible` constraints between a newtype and its underlying type aren’t solved unless the newtype constructor is exported:

```purs
module Example.N (N) where

newtype N a = N a
```

```purs
module Example where

import Example.N (N)
import Safe.Coerce (coerce)

example :: forall a. a -> N a
example = coerce
```

```
Error found:
in module Example
at Example.purs:7:11 - 7:17 (line 7, column 11 - line 7, column 17)

  No type class instance was found for
                                
    Prim.Coerce.Coercible a0    
                          (N a0)
                                

while checking that type forall (a :: Type) (b :: Type). Coercible a b => a -> b
  is at least as general as type a0 -> N a0
while checking that expression coerce
  has type a0 -> N a0
in value declaration example

where a0 is a rigid type variable
        bound at (line 7, column 11 - line 7, column 17)

See https://github.com/purescript/documentation/blob/master/errors/NoInstanceFound.md for more information,
or to contribute content related to this error.
```

Does that alleviate your concerns?

**Comment by garyb** [comment]
Ah yeah, that works too! :+1:

**Comment by garyb** [comment]
@hdgarrood I was thinking of things like, in `purescript-uri` there are a bunch of invariant-preserving newtypes over `String` that ensure only valid chars are usable (escaping isn't supported in the URI scheme, for example). It's mostly domain types rather than libraries though.

**Comment by hdgarrood** [comment]
Ah, I see, right. To clarify, the only types which need role annotations to prevent you from violating their invariants are the types which have parameters and where their representation depends on the behaviour of the instances of some of those parameters. Map’s insert depends on the Ord instance of the key, for example, so we want to prevent coercing from `Map Int a` to `Map (Down Int) a`. I don’t think there are very many types in this category. Types which don’t take parameters can’t be given role annotations anyway; roles only come into play for coercions “under” a type constructor, so like coercing from `T a b c` to `T d e f`.

**Comment by garyb** [comment]
> Types which don’t take parameters can’t be given role annotations anyway

Of course! I was definitely overlooking that aspect.

**Comment by joneshf** [comment]
@hdgarrood I feel like you're getting hung up on the time-based releases and not hearing the rest of my message. I'm not arguing that we need to be more dogmatic about time-based releases. I'm arguing that we haven't addressed the actual problem by making this release. I'm not commenting here to argue about time-based releases. As mentioned:

> If we don't want to do time-based releases anymore, that's fine.

I'm not saying time-based releases are the only way to address the problems. I offered it up because it can be a forcing function for solving release and process problems, if you're willing to be dogmatic about it. We're clearly not willing to be dogmatic about it, because we took 18 weeks to make the last release. Doing it halfway puts us in the situation we're in where it takes someone forking, making releases, and blogging about it for us to do what we should have already done.

This is the part I want you to hear: if someone in the community has to get loud about releases, they're pointing out a problem. Addressing the symptom is good. But, solving the problem is what we should actually be doing. And that person shouldn't have to be Phil in order for us to listen to them.

As you pointed out, the last big deal was that we released 0.12.0, broke a ton of stuff, and it took a long time to fix it. We addressed the feedback by making sure more stuff was up to date before releasing. That's good. That also means that with the way we develop, we end up taking more time before making a release. The change to address breakage issues is causing the current issue of indeterminate time between releases, which causes people to fork, lose faith, etc. The issue now is that we're being too dogmatic about making sure everything is up to date, and it's causing other problems.

I suggested time-based releases because it pushes to rethink processes, as long as we're willing to be dogmatic about release time. If we know that we're about to merge something that's breaking, we shouldn't say, "Let's merge this now and figure out whatever breaks later." We should alter our process to say, "Let's get everything in place **before** merging." Specifically because we know that addressing the breakage later leads to situations like we're in now. If that means we can't merge coercible or polykinds or whatever once the compiler tests pass, then that particular contribution is held up on being merged, not on being released.

One of two things would eventually come out of that process change: people will slow down making large-sweeping breaking changes, or it'll get easier to address downstream breakage. Slowing down making large-sweeping breaking changes doesn't mean changes stop being made. But those changes might start being backwards compatible, because people don't want to fix a ton of stuff. Addressing downstream breakage could mean that instead of [removing core tests](https://github.com/purescript/purescript/issues/2837#issuecomment-618577133), we keep them so a contribution has fast, scoped feedback on what needs fixing.

The important part is the process change (which we haven't addressed), not the thing that causes the process change.

**Comment by hdgarrood** [comment]
I think we need a more complete analysis of what's holding this release up before we can discuss solutions. Thinking about it a little more, I realise there are more factors which we've missed. There are two big factors which contributed to 0.14.0 taking longer which we haven't yet addressed:
- We tend to make breaking changes to the core libraries at the same time as we make breaking changes to the compiler, again to reduce the frequency with which we break things. There are a few breaking changes we might want to get in across various core libraries alongside the 0.14.0 release, which are not fully implemented/reviewed/merged:
  - https://github.com/purescript/purescript-lists/pull/161
  - https://github.com/purescript/purescript-lists/pull/150
  - https://github.com/purescript/purescript-strings/issues/91
  - https://github.com/purescript/purescript-foldable-traversable/issues/69
- Coercible _did_ appear to have everything in place at the point it was merged. It was only later that we discovered showstopping bugs such as #3870, #3871, #3876, and #3857. (I agree that fixing all known bugs before releasing is not usually desirable, but I think these particular bugs are too severe for us to be able to release without fixing them).

We mostly are making sure everything is in place by the time it is merged already. (The removal of core tests is a red herring, that script hadn't been run in years anyway).

> The issue now is that we're being too dogmatic about making sure everything is up to date, and it's causing other problems.

There's an unstated value judgment behind this conclusion that I don't think I agree with. The choices we make here are a tradeoff between getting releases out quickly, reducing the amount of effort required to make changes in the first place, reducing the frequency of breaking changes, and also trying to actually make sure that proposed breaking changes actually get done without waiting too long (because the longer we wait, the more painful it is to fix them). You're advocating moving more towards the "getting releases out quickly" end, but that comes at a cost of at least one of the other three points. It's not necessarily clear that that's a good tradeoff.

**Comment by joneshf** [comment]
> You're advocating moving more towards the "getting releases out quickly" end, but that comes at a cost of at least one of the other three points. It's not necessarily clear that that's a good tradeoff.

It's incredibly frustrating to type out a long explanation of what I'm _actually_ advocating for just to have you ignore it. I think I'm done with this conversation.

**Comment by hdgarrood** [comment]
I don’t think that’s at all a fair assessment of what I’ve done, but sure.

**Comment by gabejohnson** [comment]
I'm glad you made this release @hdgarrood (the changes to externs cut our build time by 14s 🎉) and sympathize with your desire to postpone breaking releases. At the same time, I also share @joneshf's frustrations with the current process and think _some_ kind of forcing function is necessary to make the changes happen.

> I'm not saying time-based releases are the only way to address the problems. I offered it up because it can be a forcing function for solving release and process problems, if you're willing to be dogmatic about it. We're clearly not willing to be dogmatic about it, because we took 18 weeks to make the last release. Doing it halfway puts us in the situation we're in where it takes someone forking, making releases, and blogging about it for us to do what we should have already done.

Thanks for the explanation @joneshf 😄. I also initially thought you were specifically advocating for strict time-based releases.

> The approach that I thought we had agreed on was "aim for 6 weeks between releases, except for breaking releases, which may take longer."

I don't see why these two things have to be in conflict. Is there a good reason to merge breaking changes into `master` rather than a `breaking`, `next`, or `0.14.0` branch? I don't know if this addresses @joneshf's concerns, but it would allow the ecosystem to gradually build up a package set compatible with the next breaking version. This would allow testing on real world applications without holding up patch releases while still maintain a rapid release cycle.

> We tend to make breaking changes to the core libraries at the same time as we make breaking changes to the compiler, again to reduce the frequency with which we break things.

There's a difference between breaking changes to a library that accommodate breaking changes in the compiler and those that "simply" break the API. I would prefer those _not_ be coupled so I can choose to update libraries and compiler versions independently. I find it more difficult to find the source of bugs when I have to do stop-the-world updates to a codebase.

I'd personally also rather have individual breaking changes to the compiler released for the same reason. I'd much rather update a project to deal with polykinds and then later fix things for roles than to do both at the same time.

> There's an unstated value judgment behind this conclusion that I don't think I agree with. The choices we make here are a tradeoff between getting releases out quickly, reducing the amount of effort required to make changes in the first place, reducing the frequency of breaking changes, and also trying to actually make sure that proposed breaking changes actually get done without waiting too long (because the longer we wait, the more painful it is to fix them).

I'm conscious of the fact that updating the ecosystem to accommodate breaking changes in the compiler is laborious and can put a lot of burden on both compiler and library maintainers. This feels like the real problem. Is there something we can do to address this issue specifically?

Perhaps we could have scripts for source-to-source transformations? I know there are projects in other language ecosystems (Python, Elm?) that do this. This would be useful for library maintainers and application developers alike. Maybe a breaking PR could require a migration script (or addition to a migration script) to help mitigate breakage woes.

Another possibility (on the library side of things) is to use a mono-repo for core and contrib libraries. Libraries could be published separately but changes made in a single repository. I believe it's something the compiler repo has started doing with `purecript-cst` and `purescript-ast`.

@hdgarrood I know being a maintainer of such an important project/ecosystem can be a lot of (unpaid and thankless) work and very much appreciate your efforts. I also know that I speak for many in the community when I say that I'm open to helping reduce that burden in any way my time resources and competencies allow.

**Comment by natefaubion** [comment]
We have already tried to mitigate breakage as much as possible for this release. Changes for PolyKinds are warnings wherever possible (for example we preserve `#` and `foreign import kind` syntax but warn on it) with suggested fixes that can be applied automatically by tooling. Changes to `Prim` types to introduce polymorphism (such as to `RowList`) are difficult to automate away however, as there are new semantics to consider. Roles are additive and won't prevent old code from compiling, but there are new semantics to consider in your library.

I would love to have an automatic migration tool, but I don't think that's a discussion for this release necessarily, and it can be developed independently first. I don't think it's something that needs to be factored into our current release process because it doesn't exist.

**Comment by natefaubion** [comment]
I'll note that @joneshf's timline put `Coercible` in the "not actually a breaking change" category, which while technically true, has breaking effects on the ecosystem to the point that code which supports `Coercible` via role annotations will not compile with older compilers, which makes it an effective breaking release. I personally think that warrants a breaking bump in the compiler version. How much of an effect that has on the ecosystem isn't totally clear to me though.

PolyKinds and Coercible were both merged shortly after each other. I personally agree with @hdgarrood in that these are breaking enough to warrant batching their releases. Every breaking release comes with not-insignificant negative community feedback, even for things that are _widely_ thumbs-upped, and we are trying to mitigate that. Additionally, maintaining feature branches for large changes is incredibly expensive, and I think deferring releases is an acceptable trade-off. I do not agree there was any sort of "lets merge it now and see what breaks" process here. I understand where this feeling is coming from, but I also feel as though both @hdgarrood and I did our best to make sure we merged these features when they were ready to the best of our understanding. I don't think anything was rushed. Coercible and roles were in a PR for over a year and half. Oh, but why should we have PRs like that open for so long? We've received negative feedback in that regard many times over! Perhaps any PR that sits open for too long we should just merge anyway! We could have merged and released both of these features ASAP, and we would have released an unsound feature and performance regressions. Now, what sort of pushback would we have gotten if _that_ had been the case? I feel like we can't win here.

> Doing it halfway puts us in the situation we're in where it takes someone forking, making releases, and blogging about it for us to do what we should have already done.

I think this puts the events in a far more negative light than necessary. The commits that largely prompted a community response were significant performance improvements, which was merged well after both of the large breaking features. The span of time between someone(s):

* Expressed thanks for the performance improvements.
* Suggested a 0.13.7 backport release.
* Released backport changes of their own.
* Made an official backport release.

Was a matter of a week and a half. Note that the original author of the backport changes never brought it up as a request in any normal community channels. I personally don't think there was any negative intent behind, and the blog post did not express frustration with feeling like they had to undergo such a task. I feel like that time frame is very responsive, all things considered. Someone expressed a community need, and it was addressed as best as we could.

We haven't really had features of this scale since trying more frequent releases, so this is an important learning exercise. My personal take-away is that last-resort community accountability is very problematic, but I don't see a clear "failing" here. I think there is a place for release coordination outside of the current core maintainers. If someone wanted to take on the responsibility of monitoring release progress and engaging with the community in an ongoing way over that, I think that would be _incredibly_ helpful.

**Comment by hdgarrood** [comment]
> There's a difference between breaking changes to a library that accommodate breaking changes  in the compiler and those that "simply" break the API. I would prefer those not be coupled so I can choose to update libraries and compiler versions independently. I find it more difficult to find the source of bugs when I have to do stop-the-world updates to a codebase.

I completely understand this viewpoint, and am open to this (especially since I think it would go a long way to getting breaking compiler releases out more smoothly and with fewer delays). I'd want to get feedback from the wider community on this before changing our approach, though.

> I'd personally also rather have individual breaking changes to the compiler released for the same reason. I'd much rather update a project to deal with polykinds and then later fix things for roles than to do both at the same time.

I'm less enthusiastic about this idea, personally. I can see that this might be useful for a project maintainer, but from the perspective of a library author, I think having breaking releases more often creates more fuss and risks splitting the ecosystem more.

> Is there a good reason to merge breaking changes into master rather than a breaking, next, or 0.14.0 branch?

If we did that, that probably would fix the problem of not being able to release in times like the one we're in now, but the cost is that ongoing development becomes quite a bit more expensive as we need to keep merging master into the 'next' branch too. For many changes that's surmountable, for others it could be really painful. I expect polykinds would have been an example of the latter, for instance.

> Another possibility (on the library side of things) is to use a mono-repo for core and contrib libraries. Libraries could be published separately but changes made in a single repository. I believe it's something the compiler repo has started doing with purecript-cst and purescript-ast.

Unfortunately this won't be possible until the new registry is further along, as there's an assumption built into most of the package management / build tools right now that one repo = one package. Unless we want to merge all of core into a single package (which is also something I personally am open to discussing but might be too much effort to be worth it at this stage).
